moodle/Sniffs/PHPUnit/ParentSetUpTearDownSniff.php

```diff
@@ -0,0 +1,222 @@
+                continue;
+            }
+
+            // Iterate over all the methods in the class.
+            $mStart = $cStart;
+            while ($mStart = $phpcsFile->findNext(T_FUNCTION, $mStart + 1, $tokens[$cStart]['scope_closer'])) {
+                $method = $phpcsFile->getDeclarationName($mStart);
+                // Only interested in setUp and tearDown methods.
+                if (!in_array($method, $allMethods)) {
+                    continue;
+                }
+
+                // Iterate over all the parent:: calls in the method.
+                $pStart = $mStart;
+                $correctParentCalls = [];
+                while ($pStart = $phpcsFile->findNext(T_PARENT, $pStart + 1, $tokens[$mStart]['scope_closer'])) {
+                    // The next-next token should be the parent method being named.
+                    $methodCall = $phpcsFile->findNext(T_STRING, $pStart + 1, $pStart + 3);
+                    // If we are calling to an incorrect parent method, report it. No fixable.
+                    if (
+                        $methodCall !== false &&
+                        $tokens[$methodCall]['content'] !== $method &&
+                        in_array($tokens[$methodCall]['content'], $allMethods) // Other parent calls may be correct.
+                    ) {
+                        $wrongMethod = $tokens[$methodCall]['content'];
+                        // We are calling to incorrect parent method.
+                        $phpcsFile->addError(
+                            'The %s() method in unit tests cannot call to parent::%s().',
+                            $pStart,
+                            'Incorrect' . ucfirst($method),
+                            [$method, $wrongMethod]
+                        );
+                    }
+
+                    // If we are calling to the correct parent method, annotate it.
+                    if (
+                        $methodCall !== false &&
+                        $tokens[$methodCall]['content'] === $method
+                    ) {
+                        $correctParentCalls[] = $pStart;
+                    }
+                }
+
+                // If there are multiple calls to correct parent, report it. Not fixable.
+                if (count($correctParentCalls) > 1) {
+                    $phpcsFile->addError(
+                        'The %s() method in unit tests must call to parent::%s() only once.',
+                        end($correctParentCalls),
+                        'Multiple' . ucfirst($method),
+                        [$method, $method]
+                    );
+                }
+
+                // If there are no calls to correct parent, report it. Fixable.
+                if (count($correctParentCalls) === 0) {
+                    // Any weird case where the method is empty, we need at very least 1 line. Skip.
+                    $startLine = $tokens[$tokens[$mStart]['scope_opener']]['line'];
```

Comment:
I'm not sure I agree that this is the right approach.

If the method is empty shouldn't we warn about it because it will block our future reset. I feel the correct fix is instead to remove these empty methods.

Reply:
The same applies to multi-line empty methods.

Reply:
Well, my assumption was that it makes no sense to have empty setup/teardown methods, no matter if there are sameline or multiline. Why would anybody ever do that?

I explicitly prevented those having same start and end line, that, again, is extremely unlikely case, aka:

```
public function setUp() { }
```

Because the fixer implementation does not work with them. As simple as that. So I decided that it was better to skip them completely than to report and fix them.

If you think that it's worth adding a new errorcode for them, I'm happy doing so. As said, I just imagined that they were not going to happen "ever" (near) and to skip them was enough.

About multiline-empty methods, really it's similar, who is going to write an empty multiline method? And, in case it does... what do we consider empty? Do we accept comment contents, for example? 

Again, it can be done (if the method only has whitespace and comments, suggest to remove it completely). Although current approach, that is to add the call to parent... is also correct IMO.